jetty-slf4j-impl/src/main/java/org/eclipse/jetty/logging/JettyLoggerFactory.java

```diff
@@ -0,0 +1,250 @@
+//
+// This program and the accompanying materials are made available under
+// the terms of the Eclipse Public License 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0
+//
+// This Source Code may also be made available under the following
+// Secondary Licenses when the conditions for such availability set
+// forth in the Eclipse Public License, v. 2.0 are satisfied:
+// the Apache License v2.0 which is available at
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.logging;
+
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.function.Consumer;
+
+import org.slf4j.ILoggerFactory;
+import org.slf4j.Logger;
+
+public class JettyLoggerFactory implements ILoggerFactory
+{
+    protected static JettyLoggerFactory getLoggerFactory()
+    {
+        if (instance == null)
+        {
+            instance = new JettyLoggerFactory();
+        }
+
+        return instance;
+    }
+
+    protected static void setInstance(JettyLoggerFactory loggerFactory)
+    {
+        if (loggerFactory != null && instance != null)
+        {
+            System.err.printf("Replacing main Instance %s@%x with %s@%x",
+                instance.getClass().getName(),
+                instance.hashCode(),
+                loggerFactory.getClass().getName(),
+                loggerFactory.hashCode());
+        }
+        instance = loggerFactory;
+    }
+
+    private static JettyLoggerFactory instance;
+
+    private static final String ROOT_LOGGER_NAME = "";
+    private boolean initialized = false;
+    private JettyLoggerConfiguration configuration;
+    private JettyLogger rootLogger;
+    private ConcurrentMap<String, JettyLogger> loggerMap;
+
+    private JettyLoggerFactory()
+    {
+    }
+
+    public JettyLoggerFactory initialize(JettyLoggerConfiguration config)
+    {
+        configuration = Objects.requireNonNull(config, "JettyLoggerConfiguration");
+
+        loggerMap = new ConcurrentHashMap<>();
+
+        rootLogger = new JettyLogger(ROOT_LOGGER_NAME);
+        loggerMap.put(ROOT_LOGGER_NAME, rootLogger);
+
+        rootLogger.setLevel(configuration.getLevel(ROOT_LOGGER_NAME));
+        rootLogger.setAppender(new StdErrAppender(configuration));
+
+        initialized = true;
+        return this;
+    }
+
+    private void assertInitialized()
+    {
+        if (!initialized)
+        {
+            throw new IllegalStateException(this.getClass().getSimpleName() + " is not initialized yet");
+        }
+    }
+
+    /**
+     * Get a {@link JettyLogger} instance, creating if not yet existing.
+     *
+     * @param name the name of the logger
+     * @return the JettyLogger instance
+     */
+    public JettyLogger getJettyLogger(String name)
+    {
+        assertInitialized();
```

Comment:
We want to write the code in a way that this assert is not necessary - why paying the cost for _every_ call?

Reply:
This is only and impact during initialization of a Logger itself, which should occur only once per JVM per named logger.

Reply:
Nope, that is called _every_ time you get a logger. 
Assuming that it only happens once is assuming too much. For example, CometD uses per-instance loggers and there can be thousands/millions of instances created in the lifespan of the application.

Reply:
That's just a horrible use of logging in CometD, this is why Markers and MDC and KeyValuePairs exist.

Reply:
It's not "horrible use", it's actually a feature. What I'm saying is that I don't want to bind design decision to assumptions that are wrong, so let's base on the least number of assumptions.

Reply:
initialized assert removed